### ISSUE TITLE
Inconsistent error reporting with #warning and lambdas, and overload resolution for constructor call

### ISSUE THREAD

**Issue by agocke** [dotnet/roslyn#22615]
Repro

```csharp
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace VS2017FalseCS1729BugReport
{
    class Program
    {
        static void Main()
        {
#warning bug report 1
            new BugReport("No error");
#warning bug report 2

            Action bugreport1 = new Action(() =>
            {
                new BugReport("No error");
            });

            Action bugreport2 = new Action(() =>
            {
#warning bug report 3
                new BugReport("False error");
            });

            Action bugreport3 = new Action(() =>
            {
                new BugReport("False error");
#warning bug report 4
            });
        }
    }

    class BugReport
    {
        public BugReport(string arg1, string optarg = null)
        {
        }
    }
}
```

**Comment by jcouv** [comment]
Interestingly, this repros in VS (as squiggles) and sharplab, but not as a trivial test or VS build.

I suspect this is the same issue as https://github.com/dotnet/roslyn/issues/23667. 
In that issue, a diagnostic produced during binding for the semantic model, but not for emitting. I have not yet found out how such diagnostics flow into the IDE.

```C#
        [Fact]
        [WorkItem(22615, "https://github.com/dotnet/roslyn/issues/22615")]
        public void Bug()
        {
            string source = @"
using System;

class Program
{
    static void Main()
    {
#warning bug report 1
        new BugReport(""No error"");
#warning bug report 2

        Action bugreport1 = new Action(() =>
        {
            new BugReport(""No error"");
        });

        Action bugreport2 = new Action(() =>
        {
#warning bug report 3
            new BugReport(""False error"");
        });

        Action bugreport3 = new Action(() =>
        {
            new BugReport(""False error"");
#warning bug report 4
        });
    }
}

class BugReport
{
    public BugReport(string arg1, string optarg = null)
    {
    }
}";
            var comp = CreateStandardCompilation(source);
            comp.VerifyDiagnostics(
                // (8,10): warning CS1030: #warning: 'bug report 1'
                // #warning bug report 1
                Diagnostic(ErrorCode.WRN_WarningDirective, "bug report 1").WithArguments("bug report 1").WithLocation(8, 10),
                // (10,10): warning CS1030: #warning: 'bug report 2'
                // #warning bug report 2
                Diagnostic(ErrorCode.WRN_WarningDirective, "bug report 2").WithArguments("bug report 2").WithLocation(10, 10),
                // (19,10): warning CS1030: #warning: 'bug report 3'
                // #warning bug report 3
                Diagnostic(ErrorCode.WRN_WarningDirective, "bug report 3").WithArguments("bug report 3").WithLocation(19, 10),
                // (26,10): warning CS1030: #warning: 'bug report 4'
                // #warning bug report 4
                Diagnostic(ErrorCode.WRN_WarningDirective, "bug report 4").WithArguments("bug report 4").WithLocation(26, 10)
                );
        }
```

**Comment by jcouv** [comment]
Quick update: it turns out this error is produced by an analyzer (UnboundIdentifier) which looks for possible types and constructors that a fixer could generate.
From discussion with IDE team, analyzers should produce distinct diagnostics (not re-use error codes from compiler, exposed via the compiler analyzer).

I'm preparing multiple changes:
- generally block analyzers from re-using compiler diagnostics (starting with "CS" or "BC")
- this specific analyzer (and possibly other offenders) should use distinct diagnostics
- this specific analyzer may need a fix to its logic (why is it complaining here?), which I have not investigated yet

**Comment by DorCoMaNdO** [comment]
I reported this issue through the developer community site, a note that wasn't included when it was moved to be tracked here is that it wasn't an issue with VS 2013 (I only just made the switch to VS 2017), not sure about VS 2015, but if you have access to the analyzer's code from that time you might be able to find what change caused this logic error.

**Comment by jcouv** [comment]
@DorCoMaNdO You may be referencing this issue (which came from community feedback and may be yours): https://github.com/dotnet/roslyn/issues/23667

I'll get to the bottom of this and fix both issues.

**Comment by DorCoMaNdO** [comment]
@jcouv Nope, talking about this one.
https://developercommunity.visualstudio.com/content/problem/124179/false-cs1729-error-by-intellisense-caused-by-warni.html (yeah, I know I misused "delegate" when I meant to say "lambda", the word escaped my mind at the time)

**Comment by jcouv** [comment]
Thanks for the clarification.

**Comment by jcouv** [comment]
@jinujoseph This is another issue with the UnboundIdentifiers analyzer, which I didn't know how to fix, and would like to hand off to the IDE team. I'll let you triage. Thanks

With my change (https://github.com/dotnet/roslyn/pull/23776), a distinct diagnostic is now produced.
![image](https://user-images.githubusercontent.com/12466233/34063080-ca07622a-e1a4-11e7-928f-2599feb42fa2.png)

Related issue is https://github.com/dotnet/roslyn/issues/23667

**Comment by DorCoMaNdO** [comment]
Happened to come across another instance of "IDE1008", caused by a constructor overload - List<T>(IEnumerable<T> collection).
List<T>() and List<T>(int capacity) do not reproduce this.

**Comment by CyrusNajmabadi** [comment]
IDE1008 was removed.  And i no longer repro the above as having any issues.  Closing out.
